### PR TITLE
fix: async reward computation streamed prompt groups

### PR DESCRIPTION
**Pull request by mununumu**
## Problem

The previous implementation inferred prompt groups from the size of the physical transfer batch:

```python
rewards = torch.tensor(raw_rewards, dtype=torch.float)

if rewards.shape[-1] == args.n_samples_per_prompt * args.rollout_batch_size:
    rewards = rewards.reshape(-1, args.n_samples_per_prompt)
else:
    rewards = rewards.view(-1, rewards.shape[-1])
```

For the fully asynchronous DeepEyes configuration `examples/deepeyes/run_deepeyes_qwen35_9B_async.sh`

```text
rollout_batch_size = 32 prompt groups
n_samples_per_prompt = 8
global_batch_size = 256 samples
num_iters_per_train_update = 8
```

The transfer threshold now is:

```text
256 / 8 / 8 = 4 prompt groups
```

A typical streamed transfer therefore contains:

```text
4 prompt groups × 8 samples = 32 rewards
```

The old condition evaluated as follows:

```text
32 == 32 × 8
32 == 256  # False
```

The fallback then produced:

```python
rewards.view(-1, 32)  # shape: [1, 32]
```

Instead of the correct prompt-group shape:

```text
[4, 8]
```

Consequently, all 32 rewards were normalized together.

### example

Consider four prompt groups with constant rewards:

```text
Prompt 0: [0, 0, 0, 0, 0, 0, 0, 0]
Prompt 1: [1, 1, 1, 1, 1, 1, 1, 1]
Prompt 2: [2, 2, 2, 2, 2, 2, 2, 2]
Prompt 3: [3, 3, 3, 3, 3, 3, 3, 3]
```

With correct per-prompt normalization, every centered reward should be zero because there is no preference within any prompt group:

```text
Prompt 0: 0 - 0 = 0
Prompt 1: 1 - 1 = 0
Prompt 2: 2 - 2 = 0
Prompt 3: 3 - 3 = 0
```

The old implementation instead computed one mean across all 32 rewards:

```text
global mean = (8×0 + 8×1 + 8×2 + 8×3) / 32 = 1.5
```

This produced:

```text
Prompt 0: -1.5
Prompt 1: -0.5
Prompt 2:  0.5
Prompt 3:  1.5
```

Thus, samples received non-zero advantages solely because different prompts had different reward scales.

## Fix

The new implementation uses `Sample.group_index` as the source of truth. Each prompt group is then normalized independently, and the result is written back to its original positions: The fix also makes debug subsampling select complete prompt groups rather than arbitrary rows. This preserves GRPO-style reward semantics across both full rollout batches and smaller streamed transfer batches.

### CODE REVIEW

**Comment by Yangruipis** [comment]
@mununumu  Thanks for the contribution! This is the right approach, and we have a similar change internally as well. I think we can merge this first. @NINGBENZHE  , what do you think?

**Comment by NINGBENZHE** [comment]
> @mununumu Thanks for the contribution! This is the right approach, and we have a similar change internally as well. I think we can merge this first. @NINGBENZHE , what do you think?

@mununumu Great contribution! This is a solid fix with the correct approach. Thanks!

**Review by NINGBENZHE — APPROVED** [review]
/lgtm
